### PR TITLE
fix(core): resolve rstest list through the run path's planner barrier

### LINKED ISSUE #1484: browser: `list --shard` lists tests from projects outside the current shard (multi browser project)
Split out from review of #1483.

### Problem
In multi-project browser mode, `createBrowserRuntime` builds one isolated dev server per browser project from `getBrowserProjects(context)` (every browser project), and `listBrowserTests` collects from **all** `runtime.projectServers`. Each per-project manifest enumerates tests via `import.meta.webpackContext(projectRoot, { regExp: include })`, which ignores the per-shard `testFiles` filter.

So when `rstest list --shard` omits some browser projects from `projectEntries` (the sharded set), the list output still includes those omitted projects' tests. Before the per-project-instance refactor (#1483) the single shared manifest was built from `projectEntries`, so omitted projects were never enumerated.

### Scope
Edge case only: `rstest list` + `--shard` + ≥2 browser projects. The normal `rstest run` path and single-shard/single-project list are unaffected.

### Fix options
- Minimal: in `listBrowserTests`, collect only from servers whose project is present in `projectEntries`.
- Deeper: in `createBrowserRuntime`, build servers only for projects present in `projectEntries` (also removes the "a server per browser project even when filtered" startup waste), adjusting `containerServer` derivation accordingly.
- Carry the shard `testFiles` filter into the per-project manifest so within-project shard filtering is honored too (the broadest fix).

### PR DESCRIPTION
**Pull request by fi3ework**
## Summary

Structural rewrite of the previous revision of this PR (the old patch-stack is preserved in this branch's history on GitHub before the force-push). Six successive fixes here each closed one hole in `rstest list`'s hand-rolled project resolution, and review found the next hole every time. The root cause was never any single patch: the list path decided "which projects are live" from *evidence* that config hooks had already fired (an applied-environments ledger read as a lifecycle predicate, gated by shape flags like `filesOnly`, `nodeProjects.length`, and shard state), instead of from an *ordering guarantee* that they had. Every combination of {node-only | browser-only | mixed} × {shard} × {filesOnly} × {plugin with hook | plugin without hook | no plugin} × {files | empty glob | hook-adds-files} picked its own code path, and each path had to re-answer "have the hooks run yet?" with a guess.

This revision removes the question instead of patching the guesses: **`rstest list` now resolves through `createTestPlanner` (renamed from `createRunPlanner`) — the same init barrier the run command uses — and has no project-subset, shard, or validation decision of its own.**

- Both sides' `modifyRstestConfig` hooks fire inside the planner (node via `initConfigs`, browser via the files-only discovery boot), the runnable subsets are resolved from post-hook entries with `skipEmptyProjects`, and the subsets are only reachable through the returned planner — nothing can read a plan that is still moving, so a "hooks pending?" predicate is no longer expressible.
- Browser collect loads the executor over `planner.getBrowserProjectsToRun()` with the plan's executor options: a shard-empty or file-empty browser project never boots a dev server (#1484), and `configAlreadyValidated` deduplicates browser config validation instead of reprinting every unsupported-option warning per boot.
- When the plan leaves no browser tests, `validateBrowserRunConfig` still runs per each command's own policy: `list` validates whenever the browser side is empty (an unsupported provider or core/browser version mismatch fails the list even on an empty shard), while `run` keeps its existing empty-run-only check — a run with node tests deliberately skips validating a browser side its filters left empty (pinned by `e2e/filter/related.test.ts` against an invalid provider). The divergence is now stated in both orchestrators' comments; `hasValidatedBrowserConfig()` on the planner keeps either check from re-validating after the discovery boot already did.
- Node collect consumes `planner.nodeBuild` (already prepared and `initConfigs`-ed) and the plan's node subset: no compiler, dependency preparation, or pool for a zero-entry node project (the `multi-project-config` fixture's #1363 guard made every `list` on it crash), while empty-glob projects still keep an environment long enough for a hook to add files and re-enter the plan. The browser executor load + globalSetup stage stay sequenced ahead of node collection, since the stage mutates the host `process.env` that pool workers re-read at dispatch.
- `--filesOnly` becomes a pure read of the resolved plan's entries cache.
- The shard banner is now planner-owned: resolution records counts and cannot print (`resolveShardedEntries` no longer has a logging path at all), and the planner announces exactly once after the barrier, with final post-hook counts, consuming them on print — replacing the `silentShardMessage` flag that had to be threaded through every resolve site and decided the banner's cardinality per shape.
- Deleted: `createListProjectPlanState`, `applyEnvironmentGroupsToListEntries`, and the list-only browser boot with its settlement bookkeeping (~240 lines of parallel implementation).

New pins: unit tests for the browser-only shard filter (plugin-bearing projects, one discovery boot, one executor load, no duplicate validation), hook-added files to an empty-glob node project, and a compiler-name guard proving the empty node sibling gets no compiler once hooks apply; e2e regressions for #1484 (a 1-of-4 shard lists exactly one entry and builds exactly one project) and for `list` on `multi-project-config` succeeding without touching the empty node project's compiler guard.

## Related Links

Closes https://github.com/web-infra-dev/rstest/issues/1484

## Checklist

- [x] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/core/listTests.ts [hunk 1]
```diff
@@ -473,24 +507,43 @@ const collectAllTests = async ({
...
-    const { loadBrowserExecutor } = await import('./browser/loader');
-    const executor = await loadBrowserExecutor(context, browserProjects, null, {
+    activeBrowserProjects = getShardActiveBrowserProjects({
+      browserProjects,
       shardedEntries,
-      freezeShardedEntries,
       appliedModifyRstestConfigEnvironments,
     });
+    if (activeBrowserProjects.length === 0) {
+      return;
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts]: Good catch — fixed in e5006b82: when the shard leaves no browser tests, `prepareBrowserCollection` now calls `validateBrowserRunConfig` (the same validation-only load the run path uses) before returning, so an unsupported provider or a core/browser version mismatch still fails the list.

**Comment by fi3ework** [comment]
@codex review

File: packages/core/src/core/listTests.ts [hunk 2]
```diff
@@ -344,6 +345,39 @@ const collectNodeTests = async ({
...
+  if (!shardedEntries) {
+    return browserProjects;
+  }
+
+  const hooksPending = getUserRstestConfigPluginProjects(browserProjects).some(
+    (project) =>
+      !appliedModifyRstestConfigEnvironments?.has(project.environmentName),
+  );
+  if (hooksPending) {
+    return browserProjects;
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts]: Confirmed and fixed in d9b5a45f — `initModifyRstestConfigHooks` only records environments whose callbacks actually applied, so a plugin without a `modifyRstestConfig` callback kept `hooksPending` true and disabled the filtering (verified: `list --shard=1/2` on the `multi-project-config` fixture created 2 compilers). The files-only discovery boot now marks every project it initialized as settled, and a unit test pins the scenario (`should shard-filter browser projects whose plugins register no config hooks`). Compiler count drops to 1.

File: packages/core/src/core/listTests.ts [hunk 3]
```diff
@@ -175,7 +180,28 @@ const collectNodeTests = async ({
...
+  // that hook can fire.
+  const targetProjects = (
+    await Promise.all(
+      nodeProjects.map(async (project) => ({
+        project,
+        entries: await getProjectEntries({ context, project }),
+      })),
+    )
+  )
+    .filter(({ entries }) => Object.keys(entries).length > 0)
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts]: Fixed in 0d9e3363 (this review ran against cef4d39c): the filter no longer runs before `prepareRsbuild` — the node side now initializes every project, fires `modifyRstestConfig` hooks via `initConfigs` (no compilers yet), and only then resolves the runnable subset from the post-hook file set, so a hook can still add files to an empty-glob project. Pinned by the unit test `should list node tests added by modifyRstestConfig hooks to an empty-glob project`.

File: packages/core/src/core/listTests.ts [hunk 4]
```diff
@@ -204,6 +205,36 @@ const collectNodeTests = async ({
...
+  const targetProjects = (
+    await Promise.all(
+      nodeProjects.map(async (project) => ({
+        project,
+        entries: await globTestSourceEntries(project.environmentName),
+      })),
+    )
+  )
+    .filter(({ entries }) => Object.keys(entries).length > 0)
+    .map(({ project }) => project);
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts]: Intentional, matching the run path: the planner also keeps un-planned node projects inside the Rsbuild environment set (`planner.ts` — "those still need an environment for their modifyRstestConfig hooks to fire in"), so a run with a nonempty sibling compiles the empty environment there too. Constraining the dev server to the active subset would require rebuilding the Rsbuild instance after hooks settle, and would make `list` stricter than `run` for the same config. The guard-crash scenario this PR fixes (#1363's fixture) is the all-empty node side, which now early-returns before `createRsbuildServer`.

File: packages/core/src/core/listTests.ts [hunk 5]
```diff
@@ -661,146 +513,38 @@ export async function listTests(
...
+  // An invalid browser config must fail the list whether or not the plan left
+  // a browser test to collect — the executor load inside collectBrowserTests
+  // is the only other thing that would validate it. Stricter than the run
+  // path, which skips the check when node tests exist (see the empty-run
+  // branch in `runTests.ts` for the divergence and its e2e pin).
+  if (
+    browserProjects.length &&
+    !planner.hasBrowserTestsToRun() &&
+    !planner.hasValidatedBrowserConfig()
+  ) {
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts:540]: Confirmed and fixed in 08302116: a filesOnly list is a pure plan read and never loads the executor, so the standalone check now also fires when `filesOnly` is set (unless the discovery boot already validated). Pinned by a unit test asserting exactly one validation call and zero executor loads for a filesOnly list with matched browser tests.

File: packages/core/src/core/listTests.ts [hunk 6]
```diff
@@ -344,6 +362,39 @@ const collectNodeTests = async ({
...
+  appliedModifyRstestConfigEnvironments?: Set<string>;
+}): ProjectContext[] => {
+  if (!shardedEntries) {
+    return browserProjects;
+  }
+
+  const hooksPending = getUserRstestConfigPluginProjects(browserProjects).some(
+    (project) =>
+      !appliedModifyRstestConfigEnvironments?.has(project.environmentName),
+  );
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts]: This review ran against the pre-rewrite revision — `shouldPrintShardAfterConfigHooks` and `applyBrowserFilesOnlyConfigHooks` no longer exist. Since the structural rewrite, every list resolves through `createTestPlanner`, whose files-only discovery boot runs for browser-only shapes too (it is not gated on node projects) and settles no-callback plugin projects; the real collect then loads only the shard-active subset. The exact scenario — browser-only shard where a shard-empty project carries an ordinary Rsbuild plugin without `modifyRstestConfig` — is pinned by the unit test `should shard-filter browser-only projects through the plan` (one discovery boot over both projects, one executor load for the shard-active project only).

File: packages/core/src/core/listTests.ts [hunk 7]
```diff
@@ -441,135 +404,28 @@ const collectTestFiles = async ({
...
-      close: async () => {
-        const closePromises = [nodeResult.close(), browserResult.close()];
-        await Promise.allSettled(closePromises);
-        await Promise.all(closePromises);
-      },
-    };
-  }
-
-  await prepareBrowserCollection();
+  const prepared = await prepareBrowserCollection({ context, planner });
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/listTests.ts:421]: The premise does not hold for the flow list mirrors: the non-watch run assembly also runs the browser stage before any node dependency check — `runTests.ts` loads the browser executor, inits it, and awaits `runBrowserGlobalSetupStage` before dispatching cycles, and `validateRunDependencies` only fires inside the node cycle's `ensureRunResources`. The dependency-check-before-stage ordering is watch-only (`nodeExecutorToRun.validateRunDependencies()` ahead of the stage in the watch branch, and documented under the Watch mode section of AGENTS.md): a watch session persists, so it must reject before a user setup mutates anything, while a non-watch process exits right after the failure either way. Making `list` validate node deps first would make it stricter than the non-watch run for the same config — the same class of one-sided divergence this PR just removed. If the ordering should change, it belongs to a decision covering both non-watch shapes together; keeping list aligned with the non-watch run here.
